File: packages/agentenkit/test/prune.test.ts
import { describe, expect, it } from 'bun:test';
import { join } from 'node:path';
import { setupAgentCore } from '../src/runtime.js';
import { MemoryAdminStore } from '../src/admin/memory.js';
import { MemoryBus, MemoryKv, MemoryQueue, MemoryRunStreams, MemoryStorage } from '../src/adapters/memory.js';
import { openSqlite, SqliteStorage } from '../src/adapters/sqlite.js';
import { resolveConfig } from '../src/core/types.js';
import type { Storage } from '../src/ports/storage.js';

// pruneEvents: the stream-only rows older releases left in the event table
// go, in batches; an app's own types and the record stay. The Go package
// runs the same cases under the same names (prune_test.go).

const stores: Array<[string, () => Promise<Storage>]> = [
  ['memory', async () => new MemoryStorage()],
  ['sqlite', async () => new SqliteStorage(await openSqlite(':memory:'))],
];
if (process.env.TEST_PRISMA_DB) {
  stores.push(['prisma', async () => {
    const { createRequire } = await import('node:module');
    const require = createRequire(join(import.meta.dir, '../../../examples/nextjs-app/package.json'));
    const { PrismaClient } = require('@prisma/client');
    const prisma = new PrismaClient({ datasources: { db: { url: process.env.TEST_PRISMA_DB } } });
    // Start from an empty table, so the counts are this test's.
    await prisma.agentEvent.deleteMany({});
    const { PrismaStorage } = await import('../src/adapters/prisma.js');
    return new PrismaStorage(prisma);
  }]);
}

describe('prune events', () => {
  for (const [name, open] of stores) {
    it(`old stream-only rows go in batches; the record and an app's types stay (${name})`, async () => {
      const storage = await open();
      const runtime = await setupAgentCore({
        storage, admin: new MemoryAdminStore(), bus: new MemoryBus(), queue: new MemoryQueue(), kv: new MemoryKv(),
        streams: new MemoryRunStreams(),
        resolveModel: () => ({ instance: () => ({}) as any, contextWindow: 128_000 }),
        config: resolveConfig(),
      });
      const ctx = { state: {} };
      const threads = [await storage.threads.create({}, ctx), await storage.threads.create({}, ctx)];
      // What an older release wrote: chunks and state changes beside the
      // record and an app's own event.
      for (const t of threads) {
        for (const type of ['CHUNK', 'CHUNK', 'STATE_CHANGE', 'INPUT_REQUIRED', 'INVOICE_CREATED', 'STEP_COMMITTED']) {
          await storage.events.append(t.id, { type, payload: {} }, ctx);
        }
      }

      const dry = await runtime.pruneEvents({ dryRun: true });
      expect(dry.byType).toEqual({ CHUNK: 4, STATE_CHANGE: 2, STEP_COMMITTED: 2 });
      expect(dry.total).toBe(8);
      expect(await storage.events.listSince(threads[0]!.id, -1, ctx)).toHaveLength(6); // nothing went

      const done = await runtime.pruneEvents({ batchSize: 3 });
      expect(done.total).toBe(8);
      expect(done.batches).toBeGreaterThan(2); // a batch at a time
      for (const t of threads) {
        const left = (await storage.events.listSince(t.id, -1, ctx)).map((e) => e.type);
        expect(left).toEqual(['INPUT_REQUIRED', 'INVOICE_CREATED']);
      }
      expect((await runtime.pruneEvents()).total).toBe(0); // safe to run again
    });
  }
});

File: packages/go-agentenkit/prune_test.go
package agentenkit_test

import (
	"context"
	"encoding/json"
	"os"
	"path/filepath"
	"testing"

	agentenkit "github.com/eadwinCode/agentic-kit/packages/go-agentenkit"
	"github.com/eadwinCode/agentic-kit/packages/go-agentenkit/adapters/memory"
	pgstorage "github.com/eadwinCode/agentic-kit/packages/go-agentenkit/adapters/postgres"
	"github.com/eadwinCode/agentic-kit/packages/go-agentenkit/adapters/sqlite"
	memoryadmin "github.com/eadwinCode/agentic-kit/packages/go-agentenkit/admin/memory"
	"github.com/eadwinCode/agentic-kit/packages/go-agentenkit/ports"
)

// PruneEvents: the stream-only rows older releases left in the event table
// go, in batches; an app's own types and the record stay. The TS package
// runs the same cases under the same names (test/prune.test.ts).
func TestPruneEvents(t *testing.T) {
	stores := []struct {
		name string
		open func(t *testing.T) ports.Storage
	}{
		{"memory", func(*testing.T) ports.Storage { return memory.NewStorage() }},
		{"sqlite", func(t *testing.T) ports.Storage {
			db, err := sqlite.Open(filepath.Join(t.TempDir(), "prune.sqlite"))
			must(t, err)
			t.Cleanup(func() { db.Close() })
			s, err := sqlite.New(db)
			must(t, err)
			return s
		}},
	}
	if os.Getenv("TEST_ADMIN_PG") != "" {
		stores = append(stores, struct {
			name string
			open func(t *testing.T) ports.Storage
		}{"postgres", func(t *testing.T) ports.Storage {
			db := openPostgres(t)
			t.Cleanup(func() { db.Close() })
			_, _ = db.Exec("DROP TABLE IF EXISTS pr_events, pr_messages, pr_usage, pr_threads, pr_migrations CASCADE")
			s, err := pgstorage.New(context.Background(), db, pgstorage.WithPrefix("pr_"))
			must(t, err)
			return s
		}})
	}
	for _, st := range stores {
		t.Run("old stream-only rows go in batches; the record and an app's types stay ("+st.name+")", func(t *testing.T) {
			storage := st.open(t)
			ctx, sc := context.Background(), ports.StorageContext{}
			rt, err := agentenkit.SetupAgentCore(ctx, agentenkit.RuntimeOptions{
				Storage: storage, Admin: memoryadmin.New(), Bus: memory.NewBus(), Queue: memory.NewQueue(), Kv: memory.NewKv(),
				Streams:      memory.NewRunStreams(),
				ResolveModel: func(string) (agentenkit.ResolvedModel, error) { return agentenkit.ResolvedModel{}, nil },
			})
			must(t, err)
			var threads []string
			for range 2 {
				th, err := storage.Threads().Create(ctx, ports.ThreadInit{}, sc)
				must(t, err)
				threads = append(threads, th.ID)
				// What an older release wrote: chunks and state changes beside
				// the record and an app's own event.
				for _, typ := range []string{"CHUNK", "CHUNK", "STATE_CHANGE", "INPUT_REQUIRED", "INVOICE_CREATED", "STEP_COMMITTED"} {
					_, err := storage.Events().Append(ctx, th.ID, ports.NewThreadEvent{Type: typ, Payload: json.RawMessage(`{}`)}, sc)
					must(t, err)
				}
			}

			dry, err := rt.PruneEvents(ctx, agentenkit.PruneOptions{DryRun: true})
			must(t, err)
			mustEqual(t, dry.ByType["CHUNK"], int64(4), "chunks found")
			mustEqual(t, dry.ByType["STATE_CHANGE"]+dry.ByType["STEP_COMMITTED"], int64(4), "the rest found")
			mustEqual(t, dry.Total, int64(8), "total")
			left, _ := storage.Events().ListSince(ctx, threads[0], -1, sc)
			mustEqual(t, len(left), 6, "nothing went on a dry run")

			done, err := rt.PruneEvents(ctx, agentenkit.PruneOptions{BatchSize: 3})
			must(t, err)
			mustEqual(t, done.Total, int64(8), "all of them went")
			mustEqual(t, done.Batches > 2, true, "a batch at a time")
			for _, th := range threads {
				left, _ := storage.Events().ListSince(ctx, th, -1, sc)
				var types []string
				for _, e := range left {
					types = append(types, e.Type)
				}
				mustStrings(t, types, []string{"INPUT_REQUIRED", "INVOICE_CREATED"}, "what stays")
			}
			again, err := rt.PruneEvents(ctx, agentenkit.PruneOptions{})
			must(t, err)
			mustEqual(t, again.Total, int64(0), "safe to run again")
		})
	}
}
